### PR TITLE
SDK code for shared gallery, image features and blob as source

### PR DESCRIPTION
**Pull request by kangsun-ctrl**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [x] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [x] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by grizzlytheodore** [comment]
@allenjzhang pls merge. this adds new 2020-12-01 gallery.json to sdk

File: sdk/compute/Microsoft.Azure.Management.Compute/AzSdk.RP.props [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <!--This file and it's contents are updated at build time moving or editing might result in build failure. Take due deligence while editing this file-->
   <PropertyGroup>
-    <AzureApiTag>Compute_2021-03-01;Compute_2020-12-01;Compute_2020-06-30;Compute_2020-05-01;Compute_2019-12-01;Compute_2019-11-01;Compute_2019-04-01;</AzureApiTag>
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/AzSdk.RP.props:4]: Which versions are you deleting from here? and why?
Seems like more than your Gallery version.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/GalleriesOperationsExtensions.cs [hunk 2]
```diff
@@ -150,12 +154,16 @@ public static Gallery Get(this IGalleriesOperations operations, string resourceG
...
             /// </param>
+            /// <param name='select'>
+            /// The select expression to apply on the operation. Possible values include:
+            /// 'Permissions'
+            /// </param>
             /// <param name='cancellationToken'>
             /// The cancellation token.
             /// </param>
-            public static async Task<Gallery> GetAsync(this IGalleriesOperations operations, string resourceGroupName, string galleryName, CancellationToken cancellationToken = default(CancellationToken))
+            public static async Task<Gallery> GetAsync(this IGalleriesOperations operations, string resourceGroupName, string galleryName, string select = default(string), CancellationToken cancellationToken = default(CancellationToken))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/GalleriesOperationsExtensions.cs:164]: Same customization issue as above.

File: sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/SharedGalleryTests.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+{
+    public class SharedGalleryTests : VMTestBase
+    {
+        // these needs to be decided and created, especially subscription.
+        // two gallery shared to the tenant, but only one gallery shared to the sub
+        protected const string GalleryUniqueName = "97f78232-382b-46a7-8a72-964d692c4f3f-LONGLIVEGALLERYFOJNVV";
+        protected const string GalleryImageName = "jmaesscc";
+        protected const string GalleryImageVersionName = "1.0.0";
+        
+        private string galleryAccessLocation = "eastus2euap";
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/SharedGalleryTests.cs:25]: Please use a non-Canary region for these tests.

**Review by Sandido — CHANGES_REQUESTED** [review]

File: sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/SharedGalleryTests.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+
+namespace Compute.Tests
+{
+    public class SharedGalleryTests : VMTestBase
+    {
+        // these needs to be decided and created, especially subscription.
+        // two gallery shared to the tenant, but only one gallery shared to the sub
+        protected const string GalleryUniqueName = "97f78232-382b-46a7-8a72-964d692c4f3f-LONGLIVEGALLERYFOJNVV";
+        protected const string GalleryImageName = "jmaesscc";
```
[the remark this answers is not in the document]
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/SharedGalleryTests.cs:22]: This is long live resource that we also use for our bvt. So I think my team is responsible for their maintenance already. Also, I do not see any class initialization/cleanup decorator available. Without that, resource creation/cleanup is more troublesome.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/GalleriesOperations.cs [hunk 5]
```diff
@@ -134,7 +138,7 @@ public async Task<AzureOperationResponse<Gallery>> UpdateWithHttpMessagesAsync(s
         /// <return>
         /// A response object containing the response body and response headers.
         /// </return>
-        public async Task<AzureOperationResponse<Gallery>> GetWithHttpMessagesAsync(string resourceGroupName, string galleryName, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<AzureOperationResponse<Gallery>> GetWithHttpMessagesAsync(string resourceGroupName, string galleryName, string select = default(string), Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
```
[the remark this answers is not in the document]
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/GalleriesOperations.cs:141]: done

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IGalleriesOperations.cs [hunk 6]
```diff
@@ -108,7 +112,7 @@ public partial interface IGalleriesOperations
         /// <exception cref="Microsoft.Rest.ValidationException">
         /// Thrown when a required parameter is null
         /// </exception>
-        Task<AzureOperationResponse<Gallery>> GetWithHttpMessagesAsync(string resourceGroupName, string galleryName, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken));
+        Task<AzureOperationResponse<Gallery>> GetWithHttpMessagesAsync(string resourceGroupName, string galleryName, string select = default(string), Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken));
```
[the remark this answers is not in the document]
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IGalleriesOperations.cs:115]: done

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Microsoft.Azure.Management.Compute.csproj [hunk 7]
```diff
@@ -9,18 +9,13 @@
...
     <PackageReleaseNotes>
       <![CDATA[
         This is a public release of the Azure Compute SDK. Included with this release are
-        1. Added SecurityProfile, purchasePlan properties for managed Disks.
-        2. Added burstingEnabled flag in managed Disks for paid bursting functionality.
-        3. Added Standard SSD ZRS storage account type support for managed Disks.
-        4. Added RotationToLatestKeyVersionEnabled flag in managedDisk to support rotation of keys for Disk Encryption Sets.
-        5. Added supportsHibernation flag in managed Disks to support VM hibneration functionality.
-        6. Changed PrivateEndpoint property to read-only.
+        1. This is a preview release of the Azure Compute SDK with Gallery SDK.
```
[the remark this answers is not in the document]
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Microsoft.Azure.Management.Compute.csproj]: done

File: sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/SharedGalleryTests.cs (hunk 3 above)
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/SharedGalleryTests.cs:25]: same as above

**Review by allenjzhang — APPROVED** [review]

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperations.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        /// <exception cref="ValidationException">
+        /// Thrown when a required parameter is null
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when a required parameter is null
+        /// </exception>
+        /// <return>
+        /// A response object containing the response body and response headers.
+        /// </return>
+        public async Task<AzureOperationResponse<Gallery>> GetWithHttpMessagesAsync(string resourceGroupName, string galleryName, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperations.cs:82]: you also need a method just until the 'galleryName' parameter. For all customizations, you need a method that is up to but not including the new parameter. Check this for each customization.

**Review by Sandido — CHANGES_REQUESTED** [review]

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperations.cs (hunk 8 above)
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperations.cs:111]: updated.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        /// <param name='operations'>
+        /// The operations group for this extension method.
+        /// </param>
+        /// <param name='resourceGroupName'>
+        /// The name of the resource group.
+        /// </param>
+        /// <param name='galleryName'>
+        /// The name of the Shared Image Gallery.
+        /// </param>
+        public static async Task<Gallery> GetAsync(this IGalleriesOperations operations, string resourceGroupName, string galleryName)
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs:26]: The 'this' looks like incorrect syntax, remove it.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        /// <param name='resourceGroupName'>
+        /// The name of the resource group.
+        /// </param>
+        /// <param name='galleryName'>
+        /// The name of the Shared Image Gallery.
+        /// </param>
+        /// <param name='cancellationToken'>
+        /// The cancellation token.
+        /// </param>
+        public static async Task<Gallery> GetAsync(this IGalleriesOperations operations, string resourceGroupName, string galleryName, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs:46]: The 'this' looks like incorrect syntax, remove it.

**Review by Sandido — CHANGES_REQUESTED** [review]

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs (hunk 9 above)
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs:26]: I think it is actually. VMSSOperationExtension has this too: https://github.com/kangsun-ctrl/azure-sdk-for-net/blob/master/sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/VirtualMachinesOperationsExtensions.cs

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs (hunk 10 above)
Comment (kangsun-ctrl) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations/GalleriesOperationsExtensions.cs:46]: same as above

**Review by Sandido — APPROVED** [review]
Just need to address the PR check failures before merging.

**Comment by grizzlytheodore** [comment]
@allenjzhang please merge !

**Comment by Sandido** [comment]
@markcowl , @allenjzhang , looks good to me!

**Pull request merged**